### PR TITLE
#13931: Implement index fill

### LINKED ISSUE #13931: [Feature Request] Implement index fill
**Is your feature request related to a problem? Please describe.**
Implement index fill operation following pytorch description

**Describe the solution you'd like**
I implemented index fill operation to support exactly like pytorch.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by thanhnguyen1512**
### Ticket
Closes: https://github.com/tenstorrent/tt-metal/issues/13931

### Problem description
tt-metal currently does not support the exact version of index fill operation in pytorch

### What's changed
I implemented index fill operation to support exactly like pytorch. (Having input_tensor, index_tensor, dim, value).
Added tests to cover cases:
* All dimension will be tested to fill
* Int and float will be tested accordingly with bfloat16, float32, int32 data type
* Shapes of both multiple and not multiple of 32

### Checklist
- [x] Post commit CI passes
- [x] Blackhole Post commit (if applicable) - NA
- [x] Model regression CI testing passes (if applicable) - NA
- [x] Device performance regression CI testing passes (if applicable) - NA
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/pybind11/operations/__init__.hpp
Comment (o2buzzle) [inline ttnn/cpp/pybind11/operations/__init__.hpp:137]: Please move this to an appropriate location instead of a top level module (like `data_movement`)

File: ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp
Comment (o2buzzle) [inline ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp:82]: If you are not going to support other types of data here, it is better to throw `TT_FATAL` to prevent a case of unexpected behaviors.

File: ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp
Comment (o2buzzle) [inline ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp:97]: You should use `use_dram` here instead, makes the code easier to follow

File: ttnn/cpp/pybind11/operations/__init__.hpp
Comment (thanhnguyen-moreh) [reply ttnn/cpp/pybind11/operations/__init__.hpp:137]: I think we need to discuss this with TT's code owner first

File: ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp
Comment (thanhnguyen-moreh) [reply ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp:82]: I think it'll be more reasonable to throw `TT_FATAL` in validating so I'll be doing that instead. Is that okay?

File: ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp
Comment (thanhnguyen-moreh) [reply ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp]: Resolved

File: ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp
Comment (thanhnguyen-moreh) [reply ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp]: Resolved

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (o2buzzle) [inline tests/ttnn/unit_tests/operations/test_index_fill.py:56]: You don't need this. `to_torch` should have already moved it from the device to the CPU

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (o2buzzle) [inline tests/ttnn/unit_tests/operations/test_index_fill.py:99]: You don't need this. `to_torch` should have already moved it from the device to the CPU

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (thanhnguyen-moreh) [reply tests/ttnn/unit_tests/operations/test_index_fill.py]: Resolved

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (thanhnguyen-moreh) [reply tests/ttnn/unit_tests/operations/test_index_fill.py]: Resolved

**Review by o2buzzle — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/index_fill/device/index_fill_multi_core_factory.cpp:113]:
```suggestion
    auto reader_kernel_id = CreateKernel(
        program,
```

File: ttnn/cpp/pybind11/operations/__init__.hpp
Comment (hschoi4448) [reply ttnn/cpp/pybind11/operations/__init__.hpp:142]: Could you please check this part? The rest looks good.

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (mrshaw01) [inline tests/ttnn/unit_tests/operations/test_index_fill.py:18]: Test more shapes. 2d? 3d? 5d?.

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (mrshaw01) [inline tests/ttnn/unit_tests/operations/test_index_fill.py:109]: Test more shapes. 2d? 3d? 5d?.

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (mrshaw01) [inline tests/ttnn/unit_tests/operations/test_index_fill.py:58]: Code reuse is needed for these lines. You should create something like run_index_fill and call it as needed.

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (thanhnguyen-moreh) [reply tests/ttnn/unit_tests/operations/test_index_fill.py:18]: Okay, I got it.

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (thanhnguyen-moreh) [reply tests/ttnn/unit_tests/operations/test_index_fill.py:58]: Noted, will fix it soon

Comment (thanhnguyen-moreh) [reply tests/ttnn/unit_tests/operations/test_index_fill.py]: Resolved

File: tests/ttnn/unit_tests/operations/test_index_fill.py
Comment (thanhnguyen-moreh) [reply tests/ttnn/unit_tests/operations/test_index_fill.py]: I added 2 more cases for 3D (If I were to add one more type of dimension, the number of tests might be too much).

**Review by mrshaw01 — APPROVED** [review]

**Comment by ntarafdar** [comment]
@yugi957 can you have a look at this. I think you were looking at Fill. Might not be exactly the same but can be useful.

File: ttnn/cpp/ttnn/operations/index_fill/index_fill_pybind.cpp
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/index_fill/index_fill_pybind.cpp:18]: maybe out of the scope of this but can we look at in the future supporting the square bracket notation on the left hand side.

e.g x[:2] = y

is this something we should do at a higher level @ayerofieiev-tt ?

File: ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:69]: use shift instead of division, division is expensive on the risc-v cores

File: ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:84]: same division comment as above

File: ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp
Comment (thanhnguyen-moreh) [reply ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:69]: Is it better if I were to pass in the pre-computed variable from the host code?

**Review by dmakoviichuk-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp
Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:69]: There's a trade off when moving args to compile time vs. keeping them as runtime. In this case, you could just move `index_page_size` to compile time and do this inside kernels:
```
constexpr uint32_t index_page_size_divided_by_four = index_page_size << 2;
```

Comment (TT-BrianLiu) [reply ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:69]: Also, please don't hard-code non-obvious constants like this. Give it a proper name

Comment (thanhnguyen-moreh) [reply ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:69]: Oh my bad, I'm gonna fix it rn

Comment (thanhnguyen-moreh) [reply ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp:69]: This is actually the index_size (number of elements) that I already passed as runtime. Somehow the code got reverted :/

**Review by TT-BrianLiu — APPROVED** [review]
Similar to: https://github.com/tenstorrent/tt-metal/pull/13524#discussion_r1811003165

File: ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp
Comment (hschoi4448) [reply ttnn/cpp/ttnn/operations/index_fill/device/kernels/reader_index_fill.cpp]: I was under the impression that the TT compiler would support optimization for constant division. Does the TT compiler not have an optimization feature?
@TT-BrianLiu

**Comment by thanhnguyen-moreh** [comment]
Are there any other changes that I need to make before merging? @TT-BrianLiu

**Comment by ntarafdar** [comment]
@yugi957 FYI

**Pull request merged**
